### PR TITLE
Shibboleth: Remote Authentication Phase 1

### PR DESCRIPTION
**Pull request by pdurbin**
# RFI Checklist

### 1. Related Issues

- #2939 Shibboleth: Remote Authentication Phase 1 (parent issue for [2939-shib](https://github.com/IQSS/dataverse/tree/2939-shib) branch)
- #1533 Shibboleth: Re-test Institution-wide Shibboleth groups and CRUD API operations
- #1706 Account Information API Token - Return to Account Pg? 
- #2046 Review what shows in user account info and what can be changed (Shib users can't change anything)
- #2104 Shibboleth: register with InCommon after updating entityID
- #2481 API Tokens: Add Done Button To Page
- #2512 Shibboleth: multiple email addresses are stored as a single one
- #2838 Local account creation URL ignores :AllowSignup
- #2915 Shibboleth: Add superuser API endpoint to migrate Shib user to builtin/local account
- #2916 Shibboleth: tooling for debugging SAML assertions
- #2937 Shibboleth: Support "Federated Login Mode" (i.e. feed of Identity Providers from InCommon)
- #2950 Shibboleth: Review Text For Converting an Account/Creating an Account (the new #796, related to #1497 Shibboleth: persist institution name as affiliation for users)
- #2951 Shibboleth: Institution Log In changes: Help link and extra text (the new #794)
- #2952 Shibboleth: successful conversion from builtin/local account message should appear on account page rather than the home page
- #2953 Installation Guide: Improve Shibboleth documentation
- #2966 Notifications: Welcome to 4.0 Notification Needs to Change
- #2975 User Guide: Add section about remote authentication
- #3009 API: add endpoint for listing users
- #3011 Shibboleth: Update any mentions of Institution to be Institutional
- (no issue) 404 and 403 error message text should now link to support pop up. Done in 51d15f8
- (no issue) review dataverseuser.xhtml and see if code related to SearchIncludeFragment can be removed since MyData probably isn't using it. Done in 040a4d4
- #3036 Support Form: math challenge doesn't work from some pages
- #3037 Shibboleth: Add shibboleth api endpoints to api doc
- #3041 My Data: Permissions granted by Shib groups do not appear on My Data
- #3045 Shibboleth: Convert account page fails due to password validation.
- #3080 Shib users do not get the "Welcome to Dataverse" message when creating a new account
- #3085 Shibboleth UI Improvements: Messaging, Formatting, Etc
- #3086 API Token: Add as a tab in the my data/notifications/account info page
- #3091 Shibboleth: "Allow me to type the name of my institution" option dropdown looks bad

---

### 2. Pull Request Checklist

- [x]  Functionality completed as described in [Remote Authentication Business Requirements Document](https://docs.google.com/document/d/1vcAmo2nkFYavAr7OwwXzxM0IFQbkRZYZrrX43q-wqGE/edit?usp=sharing) (BRD)
- [x]  Dependencies, risks, assumptions in [Remote Authentication Business Requirements Document](https://docs.google.com/document/d/1vcAmo2nkFYavAr7OwwXzxM0IFQbkRZYZrrX43q-wqGE/edit?usp=sharing) (BRD) addressed
- [x]  Unit tests completed
- [x]  Deployment requirements identified (e.g., SQL scripts, indexing)
- [x]  Documentation completed
- [x]  All code checkins completed

---
### 3. Review Checklist

_**After** the pull request has been submitted, fill out this section._

- [ ]  Code review completed or waived
- [ ]  Testing requirements completed
- [ ]  Usability testing completed or waived
- [ ]  Support testing completed or waived
- [ ]  Merged with develop branch and resolved conflicts

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/RoleAssigneeServiceBean.java [hunk 1]
```diff
@@ -268,10 +284,26 @@ private String getDvObjectIdListClause(List<Long> dvObjectIdList){
...
-    
+
+    private List<String> getUserRuntimeGroups(AuthenticatedUser au) {
+        List<String> retVal = new ArrayList();
+
+        Set<Group> groups = groupSvc.groupsFor(au, null);
+        StringBuilder sb = new StringBuilder();
+        for (Group group : groups) {
+            logger.fine("found group " + group.getIdentifier() + " with alias " + group.getAlias());
+            if (group.getGroupProvider().getGroupProviderAlias().equals("shib") || group.getGroupProvider().getGroupProviderAlias().equals("ip")) {
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/RoleAssigneeServiceBean.java]: I'm having trouble testing IP Groups.

**Comment by pdurbin** [comment]
@scolapasta passed this pull request to me the other day but I'm now passing it back to QA with the following updates:

- @raprasad did code review of @sekmiller 's MyData change for #3041
- I fixed the support link regressions in #3036 and moved that issue to QA (#2966 should be retested as well)
- I cleaned up the identity federation docs in #2937 and moved that issue to QA.

**Comment by kcondon** [comment]
Liz, functionality all working, did you want to take a last look on beta? Pass back to me when done, thanks.

**Comment by eaquigley** [comment]
Good on my end. Passing to @kcondon to merge.

**Comment by pdurbin** [comment]
The merging of several pull requests yesterday has resulted in the message, "This branch has conflicts that must be resolved." I'm stealing this pull request back to resolve merge conflicts.

**Comment by pdurbin** [comment]
@scolapasta it seems that merging #3108 (3089-jsoup) into develop cause several merge conflicts for this pull request. I just merged "develop" into 2939-shib (this pull request) and resolved conflicts in fe5b00d. That commit message has the details of the nature of the conflict but I'll repeat them here:

Conflicts from #3108 (3089-jsoup):
    src/main/java/Bundle.properties
    src/main/webapp/dataverseuser.xhtml
    src/test/java/edu/harvard/iq/dataverse/util/BundleUtilTest.java

- Had to fix welcome notification on signup (no more
  notification.welcome1 and notification.welcome2)
- Apply dataverse.saved.search.success fix from 09454b2

@scolapasta I'm going to pass this pull request to you to see what you think. If you could also please let me know if you think I should to the refactoring and cleanup I mentioned the other day, I'd appreciate it. I'm feeling a bit guilty about about the state of the code in the "convertShibToBuiltIn" method I added to the API: https://github.com/IQSS/dataverse/blob/fe5b00d0d75fbbb8c30837287cc4e835ec3126ea/src/main/java/edu/harvard/iq/dataverse/api/Admin.java#L269 . I hope the todo's there explain how the logic should be centralized (so it can be called from the GUI someday as well) but my biggest concern is the "half converted" stuff... ideally we would write to the "builtinuser" and "authenticateduser" tables in a single (EJB?) transaction. I'm not quite sure how to do this.

**Comment by pdurbin** [comment]
@scolapasta gave me the go ahead to clean up the "convertShibToBuiltIn" code, which I did in a new pull request: #3120.

**Pull request merged**